include/swift/IDE/CodeCompletionResultType.h

```diff
@@ -107,54 +109,202 @@ enum class CodeCompletionResultTypeRelation : uint8_t {
+  /// supertypes.
+  static const USRBasedType *null(USRBasedTypeArena &Arena);
+
+  /// Construct a \c USRBasedType from an AST-bound type. Computes the type's
+  /// supertypes.
+  static const USRBasedType *fromType(Type Ty, USRBasedTypeArena &Arena);
+
+  /// Construct as \c USRBasedType from a USR and its supertypes. This method
+  /// takes care of copying \p USR and \p Supertypes to \p Arena.
+  static const USRBasedType *fromUSR(StringRef USR,
+                                     ArrayRef<const USRBasedType *> Supertypes,
+                                     USRBasedTypeArena &Arena);
+
+  StringRef getUSR() const { return USR; }
+  ArrayRef<const USRBasedType *> getSupertypes() const { return Supertypes; }
+
+  /// Compute the relation of \c ResultType when to this contextual type.
+  CodeCompletionResultTypeRelation
+  typeRelation(const USRBasedType *ResultType,
+               const USRBasedType *VoidType) const;
+};
+
 /// The type returned by a \c ContextFreeCodeCompletionResult. Can be either of
 /// the following:
-///  - Have the NotApplicable flag set: The completion result doesn't produce
-///    something that's valid inside an expression like a keyword
-///  - An null type if the completion result produces something that's
+///  - NotApplicable: The completion result doesn't produce something that's
+///    valid inside an expression like a keyword
+///  - An empty list of types if the completion result produces something that's
 ///    valid inside an expression but the result type isn't known
-///  - A proper type if the type produced by this completion result is known
+///  - A list of proper type if the type produced by this completion result is
+///    known
+/// A \c CodeCompletionResultType does not have a single unique type because for
+/// code completion we consider e.g. the expression \c Int as producing both an
+/// \c Int metatype and an \c Int instance type. Thus we consider both the
+/// instance and the metatype as result types of the expression \c Int.
 class CodeCompletionResultType {
-public:
-  enum class Flags : unsigned {
-    IsNotApplicable = 1 << 0,
-    /// If \p AlsoConsiderMetatype is set the code completion item will be
-    /// considered as producing the declared interface type (which is passed as
-    /// \p ResultTypes ) as well as the corresponding metatype.
-    /// This allows us to suggest 'Int' as 'Identical' for both of the following
-    /// functions
-    ///
-    ///   func receiveInstance(_: Int) {}
-    ///   func receiveMetatype(_: Int.Type) {}
-    AlsoConsiderMetatype = 1 << 1
-  };
+  bool IsNotApplicable;
 
-private:
-  llvm::PointerIntPair<Type, 2, OptionSet<Flags>> TypeAndFlags;
+  /// The types a code completion result can produce.
+  /// This is not a single unique type because for code completion we consider
+  /// e.g. \c Int as producing both an \c Int metatype and an \c Int instance
+  /// type.
+  SmallVector<PointerUnion<Type, const USRBasedType *>, 1> ResultTypes;
```

Comment:
I understand that using a "collection" might be a good solution.

However, again, `CodeCompletionResultType` resides in `BumpPtrAllocator` so we can't use `SmallVector` which requires the destructor call. `ArrayRef` might be OK, but how about using `llvm::TrailingObject` so we can pack the `size()` with `IsNotApplicable` flag, and save a pointer size of `data()`. `USRBasedType` could use `TrailingObject` too...

Reply:
Ah, now I finally understand what your concern about the collection was. As discussed offline, I decided to store the result types using two pointers since we don’t need a collection with more than 2 elements at the moment anyway https://github.com/apple/swift/pull/41000/commits/9ffc3a0543f5f197ccfb116646af7b9976d23bd7